gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/CallStep.java

```diff
@@ -0,0 +1,338 @@
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.Parameters;
+import org.apache.tinkerpop.gremlin.process.traversal.traverser.TraverserRequirement;
+import org.apache.tinkerpop.gremlin.process.traversal.traverser.util.DummyTraverser;
+import org.apache.tinkerpop.gremlin.process.traversal.traverser.util.TraverserSet;
+import org.apache.tinkerpop.gremlin.process.traversal.util.FastNoSuchElementException;
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalUtil;
+import org.apache.tinkerpop.gremlin.structure.Graph;
+import org.apache.tinkerpop.gremlin.structure.service.Service;
+import org.apache.tinkerpop.gremlin.structure.service.ServiceRegistry;
+import org.apache.tinkerpop.gremlin.structure.util.CloseableIterator;
+import org.apache.tinkerpop.gremlin.structure.util.StringFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.tinkerpop.gremlin.structure.util.CloseableIterator.EmptyCloseableIterator;
+import static org.apache.tinkerpop.gremlin.structure.service.Service.ServiceCallContext;
+
+/**
+ * Reference implementation for service calls via the {@link ServiceRegistry} and {@link Service} APIs. This step
+ * can be used to start a traversal or it can be used mid-traversal with one at a time or barrier execution.
+ *
+ * @author Mike Personick (http://github.com/mikepersonick)
+ */
+public final class CallStep<S, E> extends AbstractStep<S, E> implements TraversalParent, AutoCloseable, Configuring {
+
+    private final boolean isStart;
+    private boolean first = true;
+
+    private ServiceCallContext ctx;
+    private String serviceName;
+    private Service<S, E> service;
+    private Map params;
+    private Traversal.Admin<S,Map> mapTraversal;
+    private Parameters parameters;
+
+    private transient Traverser.Admin<S> head = null;
+    private transient CloseableIterator iterator = EmptyCloseableIterator.instance();
+
+    public CallStep(final Traversal.Admin traversal, final boolean isStart) {
+        this(traversal, isStart, null);
+    }
+
+    public CallStep(final Traversal.Admin traversal, final boolean isStart, final String service) {
+        this(traversal, isStart, service, null);
+    }
+
+    public CallStep(final Traversal.Admin traversal, final boolean isStart, final String service, final Map params) {
+        this(traversal, isStart, service, params, null);
+    }
+
+    public CallStep(final Traversal.Admin traversal, final boolean isStart, final String service, final Map params,
+                    final Traversal.Admin<S, Map> mapTraversal) {
+        super(traversal);
+
+        this.isStart = isStart;
+        this.serviceName = service;
+        this.params = params == null ? new LinkedHashMap() : params;
```

Comment:
minor thing i suppose, but while digging through this class, i sorta kept confusing `params` with our typical usage of `parameters` - maybe we could rename `params` to something else that's more descriptive in relation to `call()`?